### DISCUSSION TITLE
Unnecessary comparison for StrEnum, how to fix?

### DISCUSSION THREAD

**Discussion by rafrafek** [microsoft/pyright#7230] [Q&A] [answered]
Hello,

I'm using the latest Pylance in VS Code in strict type checking mode and the following code gives me "unnecessary comparison" reports. How can I fix my code? Or is it just a false positive report and I can ignore it? 

I'm on Python 3.12.1.
Pylance language server 2023.12.1 (pyright a78e2b10).

```python
from enum import StrEnum

class JobStatus(StrEnum):
    QUEUED = "queued"
    PROGRESS = "progress"
    PUBLISH = "publish"
    COMPLETED = "completed"
    CANCELLED = "cancelled"
    FAILED = "failed"
    UNKNOWN = "unknown"

class WorkflowStatus(StrEnum):
    RUNNING = "RUNNING"
    SUCCEEDED = "SUCCEEDED"
    FAILED = "FAILED"

    def to_job_status(self) -> JobStatus:
        # Condition will always evaluate to False since the types
        # "Literal['RUNNING', 'SUCCEEDED', 'FAILED']" and
        # "Literal[WorkflowStatus.RUNNING]" have no overlap
        # Pylance(reportUnnecessaryComparison):
        if self.value == self.RUNNING:
            return JobStatus.PROGRESS
        # Same "unnecessary comparison" report here:
        elif self.value == self.SUCCEEDED:
            return JobStatus.COMPLETED
        else:
            return JobStatus.FAILED

print(
    WorkflowStatus.RUNNING.to_job_status(),
    WorkflowStatus.SUCCEEDED.to_job_status(),
    WorkflowStatus.FAILED.to_job_status(),
)  # Output: progress completed failed
```

I also tested with Pyright 1.1.350 with the same result:

```shell
(.venv) ➜  ... pyright enums.py
.../enums.py
  .../enums.py:24:12 - error: Condition will always evaluate to False since the types "Literal['RUNNING', 'SUCCEEDED', 'FAILED']" and "Literal[WorkflowStatus.RUNNING]" have no overlap (reportUnnecessaryComparison)
  .../enums.py:27:14 - error: Condition will always evaluate to False since the types "Literal['RUNNING', 'SUCCEEDED', 'FAILED']" and "Literal[WorkflowStatus.SUCCEEDED]" have no overlap (reportUnnecessaryComparison)
2 errors, 0 warnings, 0 informations 
(.venv) ➜  ... pyright --version
pyright 1.1.350
(.venv) ➜  ... python --version
Python 3.12.1
```

My `pyrightconfig.json` file:

```json
{
    "typeCheckingMode": "strict"
}
```

**Comment by erictraut** [comment, marked as answer]
`reportUnncessaryComparison` is a strict-mode check. Its behavior relies on some assumptions that may not always be true, as in your code sample. In theory, any Python object can override `__eq__` and `__ne__` and handle comparison any way it wants, so `reportUnnecessaryComparison` has the potential for false positives.

You have a couple of options here:
1. Modify your code so it doesn't violate the assumptions of this check. You can do so by changing `self.value == self.RUNNING` to `self.value == self.RUNNING.value`.
2. Do not enable the `reportUnnecessaryComparison` check.
3. Use `# type: ignore` or `# pyright: ignore` to suppress the error on specific lines.
4. Use `# pyright: reportUnnecessaryComparison=false` to suppress the error in specific files.

**Reply by rafrafek** [reply]
Thanks!
